File: SortColor.java
// Time Complexity : O(n) where 'n' is the length of nums.
// Space Complexity : O(1).
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

//Approach:
// maintain 3 pointers, l =i+1 and high = last index
// if we find 2, we put it at the end and decrement high
//if we find 0 we put it at the start and increment low and i
//if we find 1 we do not swap, just move to the next element
class SortColor {
    public void sortColors(int[] nums) {
        int i = 0;
        int l = 0;
        int h = nums.length - 1;
        if(nums==null || nums.length==0) return;
        while(l<=h){
            if(nums[l] == 0){
                swap(nums,i,l);
                i++;
                l++;
            }else if(nums[l] == 2){
                swap(nums,l,h);
                h--;
            }else{
                l++;
            }
        }
    }
    private void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
        // nums[i] = nums[i] + nums[j];
        // nums[j] = nums[i] - nums[j];
        // nums[i] = nums[i] - nums[j];
    }
}

File: sortcolor.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
use 3 pointers l m =0 and h = last index
check while m <=h
if m has 2 value the swap with h and reduce h
if m is 1 just increment m
else swap m and l element and  increment both
*/

package main

import "fmt"

func sortColors(nums []int)  {
	fmt.Println(nums)
	l, m, h := 0, 0, len(nums) - 1
	for m <= h {
		if nums[m] == 2 {
			temp := nums[m]
			nums[m]= nums[h]
			nums[h]= temp
			h--
		} else if nums[m] == 1 {
			m++
		} else {
			temp := nums[m]
			nums[m]= nums[l]
			nums[l]= temp
			l++
			m++
		}
	}
	fmt.Println(nums)
}

func MainSortColor() {
	sortColors([]int {2,0,2,1,1,0}) //expected  [0,0,1,1,2,2])
}
